scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DataFileService.java

```diff
@@ -0,0 +1,105 @@
+ */
+
+@Transactional
+@Service
+public class DataFileService implements IDataFileService {
+
+    private final DataFileRepository dataFileRepository;
+
+    private final DatasetRepository datasetRepository;
+    private final IFileSystemService fileSystemService;
+
+    @Autowired
+    public DataFileService(DataFileRepository dataFileRepository,
+                           DatasetRepository datasetRepository,
+                           IFileSystemService fileSystemService) {
+        this.dataFileRepository = dataFileRepository;
+        this.datasetRepository = datasetRepository;
+        this.fileSystemService = fileSystemService;
+    }
+
+    /**
+     * Creates a file on the filesystem and stores the DataFile
+     * in the database. A Dataset has many DataFiles and a DataFile belongs to one Dataset.
+     *
+     * @param bytes     The byte array representation of a MultipartFile
+     * @param datasetId The primary key of the Dataset as a String.
+     * @param username  The user / owner of the Dataset
+     * @param filename  The name of the file
+     * @param type      The file type
+     */
+    @Override
+    public DataFile createFile(byte[] bytes, String datasetId, String username, String filename,
+                               String type) throws DataFileStorageException {
+
+        DataFile newFile = new DataFile();
+        newFile.setName(filename);
+        newFile.setType(type);
+        newFile.setCreatedAt(OffsetDateTime.now());
+        Optional<Dataset> dataset;
+        Long dataFileId;
+        try {
+            dataset = datasetRepository.findById(Long.parseLong(datasetId));
+        } catch (NumberFormatException e) {
+            throw new DataFileStorageException("Dataset Id could not be parsed to a long", e);
+        }
+        if (dataset.isPresent()) {
+            newFile.setDataset(dataset.get());
+            dataFileId = dataset.get().addFile(dataFileRepository.save(newFile));
+            datasetRepository.save(dataset.get());
+        } else {
+            throw new DataFileStorageException("Dataset not found");
+        }
+
+        // Save file with unique name (its id) and maintain the file type extension
+        String indexBasedFilename = dataFileId + filename.substring(filename.lastIndexOf("."));
+        try {
+            fileSystemService.createFileInDirectory(username,
+                    Dataset.class.getSimpleName().toLowerCase(Locale.ROOT), datasetId,
+                    indexBasedFilename, bytes);
+        } catch (FileSystemStorageException e) {
+            // Saving the DataFile in the filesystem failed.
+            // Make sure the file just saved in the database doesn't remain in the database
+            dataset.get().getFiles().remove(newFile);
+            dataFileRepository.deleteById(dataFileId);
+            datasetRepository.save(dataset.get());
+            throw new DataFileStorageException("File could not be saved in the file system", e);
```

Comment:
if you use `@Transactional(rollbackFor=Exception.class)`  and just throw the exception here, you should not have to revert these changes yourself. Spring will make sure to rollback your changes. If you do it by hand, this has to be kept in mind by future developers if they make any other changes to the data in the database.